### PR TITLE
THRIFT-3783: python code generator dose not handle struct dependent

### PR DESCRIPTION
**Pull request by collinmsn**

### CODE REVIEW

File: compiler/cpp/src/generate/t_py_generator.cc [hunk 1]
```diff
@@ -21,12 +21,19 @@
...
 
 #include <stdlib.h>
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <sstream>
 #include <algorithm>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/topological_sort.hpp>
+#include <boost/graph/depth_first_search.hpp>
+#include <boost/graph/visitors.hpp>
```
Comment (Jens-G) [inline compiler/cpp/src/generate/t_py_generator.cc]: We don't want to add boost as a dependency when building the Thrift compiler. Sorry.

File: compiler/cpp/src/generate/t_py_generator.cc [hunk 2]
```diff
@@ -384,6 +406,55 @@ void t_py_generator::init_generator() {
...
     py_imports() << endl <<
     "from .ttypes import *" << endl;
+
+  construct_object_dependent_graph();
+}
+
+/**
+ * Construct object dependent graph to determine object generate order
+ * if cycle dependent is detected, generate process will be aborted
+ * object dependent example
```
Comment (Jens-G) [inline compiler/cpp/src/generate/t_py_generator.cc]: But your test case is perfectly valid Thrift IDL:
```
struct A {
  1: B b;
}

struct B {
  1: A a;
}
```

One can do this with Thrift. Is there no way to support that in Python? Something like forward-declaration of types maybe?

File: compiler/cpp/src/generate/t_py_generator.cc [hunk 3]
```diff
@@ -384,6 +406,55 @@ void t_py_generator::init_generator() {
...
+    }
+  }
+
+  DependentGraph dependent_graph(edges.begin(), edges.end(), objects.size());
+  // check cycle dependent
+  bool has_cycle = false;
+  cycle_detector detector(has_cycle);
+  boost::depth_first_search(dependent_graph, boost::visitor(detector));
+  if (has_cycle) {
+    throw "type error: cycel dependent found in " + program_->get_path();
```
Comment (Jens-G) [inline compiler/cpp/src/generate/t_py_generator.cc]: Would be nice to know which type(s) are affected.

File: test/py/object_gen_order/cycle.thrift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace py cycle
+
+struct A {
+  1: B b;
+}
+
+struct B {
+  1: A a;
+}
```
Comment (Jens-G) [inline test/py/object_gen_order/cycle.thrift]: As I said, that's legal code.

**Comment by jeking3** [comment]
Does this need further work or should it be closed?

**Comment by jeking3** [comment]
Closing due to 2 years of inactivity.

**Comment by KarDik465** [comment]
This problem is actual.
I have problems because of it

**Pull request closed without merging**